src/query/functions/aggregation/count_values.go

```diff
@@ -0,0 +1,268 @@
+func processBlockBucketAtColumn(
+	currentBucketBlock *bucketBlock,
+	values []float64,
+	bucket []int,
+	columnIndex int,
+) {
+	// Generate appropriate number of rows full of -1s that will later map to NaNs
+	// unless updated with valid values
+	currentColumnLength := currentBucketBlock.columnLength
+	currentBucketBlock.columns[columnIndex] = make(bucketColumn, currentColumnLength)
+	for i := 0; i < currentColumnLength; i++ {
+		ts.MemsetInt(currentBucketBlock.columns[columnIndex], -1)
+	}
+
+	countedValues := countValuesFn(values, bucket)
+	for distinctValue, count := range countedValues {
+		currentBucketColumn := currentBucketBlock.columns[columnIndex]
+		if rowIndex, seen := currentBucketBlock.indexMapping[distinctValue]; seen {
+			// This value has already been seen at rowIndex in a previous column
+			// so add the current value to the appropriate row index.
+			currentBucketColumn[rowIndex] = count
+		} else {
+			// The column index needs to be created here already
+			// Add the count to the end of the bucket column
+			currentBucketBlock.columns[columnIndex] = append(currentBucketColumn, count)
+
+			// Add the distinctValue to the indexMapping
+			currentBucketBlock.indexMapping[distinctValue] = len(currentBucketColumn)
+		}
+	}
+
+	currentBucketBlock.columnLength = len(currentBucketBlock.columns[columnIndex])
+}
+
+// Process the block
+func (n *countValuesNode) Process(ID parser.NodeID, b block.Block) error {
+	stepIter, err := b.StepIter()
+	if err != nil {
+		return err
+	}
+
+	params := n.op.params
+	meta := stepIter.Meta()
+	seriesMetas := utils.FlattenMetadata(meta, stepIter.SeriesMeta())
+	buckets, metas := utils.GroupSeries(
+		params.MatchingTags,
+		params.Without,
+		n.op.opType,
+		seriesMetas,
+	)
+
+	stepCount := stepIter.StepCount()
+	intermediateBlock := make([]bucketBlock, len(buckets))
+	for i := range intermediateBlock {
+		intermediateBlock[i].columns = make([]bucketColumn, stepCount)
+		intermediateBlock[i].indexMapping = make(map[float64]int, 10)
+	}
+
+	for columnIndex := 0; stepIter.Next(); columnIndex++ {
+		step, err := stepIter.Current()
+		if err != nil {
+			return err
+		}
+
+		values := step.Values()
+		for bucketIndex, bucket := range buckets {
+			processBlockBucketAtColumn(
+				&intermediateBlock[bucketIndex],
+				values,
+				bucket,
+				columnIndex,
+			)
+		}
+	}
+
+	numSeries := 0
+	for _, bucketBlock := range intermediateBlock {
+		numSeries += bucketBlock.columnLength
+	}
+
+	// Rebuild block metas in the expected order
+	blockMetas := make([]block.SeriesMeta, numSeries)
+	initialIndex := 0
+	for bucketIndex, bucketBlock := range intermediateBlock {
+		for k, v := range bucketBlock.indexMapping {
+			blockMetas[v+initialIndex] = block.SeriesMeta{
+				Name: n.op.OpType(),
+				Tags: metas[bucketIndex].Tags.Clone().AddTag(models.Tag{
+					Name:  n.op.params.StringParameter,
+					Value: strconv.FormatFloat(k, 'f', -1, 64),
```

Comment:
probably move this into a util function since we should use the same logic when you output from the render function.